### ISSUE TITLE
Investigate possible benefits of implementing java.util.Iterator

### ISSUE THREAD

**Issue by richardstartin** [RoaringBitmap/RoaringBitmap#363]
Often allocations of iterators and iterator like objects are not eliminated because methods constructing the iterator appear right before an expensive block of code (the iteration itself) so according to profiling data, the iterator construction method looks cold, and won't be inlined. Backported from JDK14, there is a heuristic to handle this case (see [JDK-8223504](https://bugs.openjdk.java.net/browse/JDK-8223504)) which forces inlining whenever the method responsible for creating the iterator is called `iterator()` and the return type implements `java.util.Iterator` (see [diff](http://cr.openjdk.java.net/~skuksenko/hotspot/8223504/webrev.00/src/hotspot/share/opto/bytecodeInfo.cpp.cdiff.html)). 

I would be interested to know if bitmap iteration could benefit from this JIT compiler heuristic. `Container` would need a new method `iterator()` which returns a `ShortIterator`. `ShortIterator` would implement `Iterator<Short>` but the method `Short next()` would not be used. 

A positive test case would be to verify that all `ShortIterator` allocations are eliminated during a bitmap iteration when using a recent JDK build.

**Comment by lemire** [comment]
What do we know about the current handling of our code by the most recent JDKs? I am not an expert in the JDK, but surely it can do things like escape analysis?

**Comment by richardstartin** [comment]
It does, but profiling informs the analysis. Because allocation of an iterator is usually associated with a subsequent loop, the allocation of the iterator may not be hot enough for the analysis to actually happen. The ticket I linked describes a new heuristic where any call site where a method called iterator produces a subclass of Iterator will be analysed automatically. I did an experiment, but couldn't find any benefit.
